### PR TITLE
HDDS-5604. Intermittent failure in TestPipelineClose.

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Fix intermittent test failure in TestPipelineClose#testPipelineCloseWithLogFailure. The mechanics of the test look good to me. Based on my testing the issue was an overly aggressive timeout waiting for a mockito invocation.

## What is the link to the Apache JIRA

HDDS-5604

## How was this patch tested?

Before patch, test failed after 30 runs locally. After patch, The test passed with 100 runs on CI: https://github.com/errose28/hadoop-ozone/runs/3354906907
    - The test was run on a different commit by accident, but the changes are identical. Check the small diff for that commit if unsure.

### CODE REVIEW

**Review by JacksonYao287 — COMMENTED** [review]
i have set the timeout here to a larger value too , but when i test this  with `@RepeatedTest(100)` locally, some flaky failure happens again. seems we need a deeper dive

**Comment by errose28** [comment]
@JacksonYao287 thanks for investigating. Are you using `org.junit.jupiter.api.RepeatedTest` annotation to run multiple times? When I try using this annotation, it does not initialize the test properly. I have been doing my repeated runs using Intellij test configuration locally, or with maven on CI. I will run another batch locally overnight and see if I can get any failures.

Did you see the failure in the same place as reported originally? What timeout value did you use for your testing?

**Comment by JacksonYao287** [comment]
> @JacksonYao287 thanks for investigating. Are you using `org.junit.jupiter.api.RepeatedTest` annotation to run multiple times? When I try using this annotation, it does not initialize the test properly. I have been doing my repeated runs using Intellij test configuration locally, or with maven on CI. I will run another batch locally overnight and see if I can get any failures.

i use the following annotations and do the test using Intellij
```
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.RepeatedTest;
```

> Did you see the failure in the same place as reported originally? What timeout value did you use for your testing?
yea, i see the same failure,  i set the timeout to 1000.

> EDIT: I was able to repro a failure after 138 runs (took 1.5 hours). Better than before, but not totally fixed, I agree. I'll move this to draft for now.

thanks for the work!

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java [hunk 1]
```diff
@@ -242,7 +238,7 @@ public void testPipelineCloseWithLogFailure() throws IOException {
     xceiverRatis.handleNodeLogFailure(groupId, null);
 
     // verify SCM receives a pipeline action report "immediately"
-    Mockito.verify(pipelineActionTest, Mockito.timeout(100))
+    Mockito.verify(pipelineActionTest, Mockito.timeout(500))
```
Comment (sodonnel) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java:241]: If it is still failing, I wonder does the timeout need to be even higher. In the init() method of this test class, I see:

```
  public void init() throws Exception {
    conf = new OzoneConfiguration();
    cluster = MiniOzoneCluster.newBuilder(conf).setNumDatanodes(3).build();
    conf.setTimeDuration(HddsConfigKeys.HDDS_HEARTBEAT_INTERVAL, 1000,
        TimeUnit.MILLISECONDS);
    pipelineDestroyTimeoutInMillis = 1000;
    conf.setTimeDuration(ScmConfigKeys.OZONE_SCM_PIPELINE_DESTROY_TIMEOUT,
        pipelineDestroyTimeoutInMillis, TimeUnit.MILLISECONDS);
   ...
```

Should the conf not be set before the MiniOzoneCluster.build() method is called? Could the cluster potentially start without the conf changes?


I cannot remember how this works. The DNs need to send a command to SCM that the pipeline should be closed here - does the DN do that over the DN heartbeat, or via some other mechanism? If the heartbeat is set to 1 second, then could this message take just over a second to arrive in the worse case?

Comment (sodonnel) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java:241]: In `SCMDatanodeHeartbeatDispatcher.dispatch()` it seems to deal with the heartbeats, and within it, there is logic for PIPELINE_ACTIONS commands. If the commands are all sent via the heartbeat, perhaps we need a timeout of 1500. I am not sure if something on the DN can trigger the heartbeat to be send immediately if certain commands are queued on the DN though.

Comment (iamabug) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java:241]: I notice that there is a global Timeout variable in this class, which applies to all the methods in this class.
```
 public class TestPipelineClose {

   /**
     * Set a timeout for each test.
     */
   @Rule
   public Timeout timeout = Timeout.seconds(300);
```

Not sure if this has something to do with the failures.

Comment (adoroszlai) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java:241]: > Should the conf not be set before the MiniOzoneCluster.build() method is called?

With 1s timeout it seems to work fine both with and without that change (tested 3-400 repetitions).

https://github.com/adoroszlai/hadoop-ozone/commits/HDDS-5604-repeat (`build()` moved after conf is set)
https://github.com/adoroszlai/hadoop-ozone/commits/HDDS-5604-repro-1000 (`build()` in original order)

Comment (sodonnel) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java:241]: Sounds like a timeout of 1000ms is the way to go then. 

I suspect the config should really all be setup before the mini-cluster is started to be sure the components get the correct config in time. Otherwise, the service could start before the config gets changed and start with the wrong value.

Comment (adoroszlai) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java:241]: > I suspect the config should really all be setup before the mini-cluster is started

Agree.

Comment (adoroszlai) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestPipelineClose.java]: I've updated the PR with both of these changes (setting conf and increasing timeout).

**Comment by errose28** [comment]
@adoroszlai how many times did you run the tests? I see you commented 3-400 repetitions, but looking at the links I only see 3 jobs with 10 runs each? Maybe I misread the results. @JacksonYao287 Said he was able to reproduce the failure even with a 1 second timeout. I have not yet tried this specific value myself but it looks like we need at least 200 clean runs to be sure based on my testing.

**Comment by adoroszlai** [comment]
> 3 jobs with 10 runs each?

3-4 runs * 10 iterations * `@RepeatedTest(10)`

**Comment by errose28** [comment]
Ah I missed the repeated test added in the diff. Thanks.

**Comment by errose28** [comment]
Thanks to everyone who has helped investigate this. Here is my current understanding of how this test is supposed to work:

1. Pipeline created
2. Log failure triggered on datanode
3. Datanode is expected to notify SCM of pipeline close without waiting for the heartbeat interval.
4. Test checks SCM event queue to see if the mock pipeline action handler it registered was called within the timeout.

So although the mini ozone conf being set with the heartbeat interval in the wrong order is bad, it should have no bearing on this test. I verified this by setting the heartbeat interval to 10 seconds. The test behaves the same.

By turning on trace logging for the EventQueue, we can see that even in the failure cases both the original PipelineActionHandler and the mocked PipelineActionHandler are triggered on SCM pretty quickly after the DN log fails. This gives me reasonable confidence that the code is working as expected and the error was in the test. So since the test seems to pass repeatedly with the new timeout, I am ok with the fix.

However, I do not fully understand why we had to increase the timeout to 1 second. SCM event queue fires asynchronuously and I verified the heartbeat interval is not blocking the datanode. I guess it is possible that DN execution time + loopback RPC latency + SCM execution time takes up to 1 second sometimes, although this seems kind of high since it previously finished under 100ms ~29/30 times, and under 500ms ~139/140 times.

TLDR I am reasonably confident the code is working correctly and this test fix is ok, but why the test fix was needed in the first place is still a bit of a mystery to me.

**Comment by avijayanhwx** [comment]
I see that a lot of analysis has gone into this already. I am wondering if we can change the test a little bit to help remove the rare failure. Since this is a mocked up test (explicitly adding pipeline actions handler, and waiting for capture), maybe we can remove the RPC call out of the test. If we can merely capture the heartbeat on the DN side to make sure it has the pipeline action that may be good enough. Thoughts?
